src/parser/Uritemplate.h

```diff
@@ -0,0 +1,396 @@
+                                          >
+                                        > {};
+
+            struct varname : pegtl::seq<
+                               varchar,
+                               pegtl::star<
+                                 pegtl::sor< // allow single dots - not continuous
+                                   noncontinuous_dots,
+                                   varchar
+                                 >
+                               >
+                             > {};
+
+            struct varspec : pegtl::seq<varname, pegtl::opt<mod_l4>> {};
+
+            struct variable_list : recover::list<
+                                     varspec,
+                                     pegtl::one<','>,
+                                     pegtl::sor<
+                                       expression_close,
+                                       pegtl::eof
+                                     >
+                                   > {};
+
+            /**
+             * dummy token to handle missing close bracket and avoid global failure while parsing
+             */
+            struct missing_expression_close : pegtl::seq<> {};
+
+            struct expression : pegtl::if_must<
+                                  expression_open,
+                                  pegtl::opt<operator_>,
+                                  variable_list,
+                                  pegtl::sor<
+                                    expression_close, missing_expression_close
+                                  >
+                                > {};
+
+            /**
+             *FIXME: character set allowed by RFC 5670 for literals:
+             *         %x21 / %x23-24 / %x26 / %x28-3B / %x3D / %x3F-5B
+             *      /  %x5D / %x5F / %x61-7A / %x7E / ucschar / iprivate
+             *      /  pct-encoded
+             *           ; any Unicode character except: CTL, SP,
+             *           ;  DQUOTE, "'", "%" (aside from pct-encoded),
+             *           ;  "<", ">", "\", "^", "`", "{", "|", "}"
+             *
+             *  But this collide with RFC1738 - Section 2.2 where are defined "unsafe characters"
+             *  which must be pct-encoded.
+             *
+             *  These characters are "{", "}", "|", "\", "^", "~", "[", "]", and "`".
+             *  Coded as in RFC5760:
+             *  %x7B / %x7D / %x7C / %x5C / %x5E / %x7E / %x5B / %x5D / %x60
+             *
+             *  So we modify original literal set from RFC5670 and remove them from allowed chars.
+             *
+             *  Original set is named `rfc_5670_ascii_allowed`
+             *  while modified is named `ascii_allowed`
+             *
+             *  From `ascii_allowed` are removed chars to be RFC1738 compatible
+             *  "~", "[" , "]"
+             *  %x7E / %x5B / %x5D 
+             *
+             */
+
+            struct rfc_5670_ascii_allowed : pegtl::sor<
```

Comment:
The relevant part of RFC6570 is 2.1 (literals)  refers to utf-8 and unicode. Should we avoid calling it ascii as the spec specifies the character encoding as utf-8 not ascii.

> 2.1.  Literals
> 
>    The characters outside of expressions in a URI Template string are
>    intended to be copied literally to the URI reference if the character
>    is allowed in a URI (reserved / unreserved / pct-encoded) or, if not
>    allowed, copied to the URI reference as the sequence of pct-encoded
>    triplets corresponding to that character's encoding in UTF-8
>    [RFC3629].
> 
>      literals      =  %x21 / %x23-24 / %x26 / %x28-3B / %x3D / %x3F-5B
>                    /  %x5D / %x5F / %x61-7A / %x7E / ucschar / iprivate
>                    /  pct-encoded
>                         ; any Unicode character except: CTL, SP,
>                         ;  DQUOTE, "'", "%" (aside from pct-encoded),
>                         ;  "<", ">", "\", "^", "`", "{", "|", "}"

Reply:
This is just subset from ASCII range

Full definition of literals (including utf-8) is later https://github.com/apiaryio/drafter/blob/6c4a9c5f20f09c18b275338188f0ab685fb3cdf3/src/parser/Uritemplate.h#L197-L204